src/main.rs

```diff
@@ -1,145 +1,136 @@
-                    process::exit(1);
-                }
-
-                match process::Command::new(&health_cmd).status() {
-                    Ok(status) => {
-                        if status.success() {
-                            let message = [("WATCHDOG", "1")];
-
-                            if let Err(err) = daemon::pid_notify(pid, false, message.iter()) {
-                                println!("{}", err);
-                                process::exit(1);
-                            };
-                        }
-                    }
-                    Err(err) => {
-                        println!("{}", err);
-                        process::exit(1);
-                    }
-                }
-
-                thread::sleep(interval);
+    if let Some(pid) = args.pid {
+        let interval = if let Some(interval) = args.interval {
+            interval
+        } else {
+            println!("WATCHDOG_USEC not set");
+            println!("Exiting, will not pet the watchdog");
+            return Ok(());
+        };
+
+        loop {
+            if kill(Pid::from_raw(pid), None).is_err() {
+                println!("Parent process exited");
+                return Ok(());
             }
-        }
-        None => {
-            let pid = getpid();
-
-            // first we start the helper child process that will run the healthcheck
-            let helper = process::Command::new("/proc/self/exe")
-                .args(&["--healthcheck", &health_cmd])
-                .args(&["--pid", pid.to_string().as_str()])
-                .spawn();
-
-            match helper {
-                Ok(mut helper) => {
-                    // Then we execve to the requested program
-                    let err = exec::Command::new(&matches.free[0])
-                        .args(&matches.free[1..])
-                        .exec();
-
-                    // We only reach this point if execve failed
-                    println!("Error: {}", err);
-
-                    helper.kill().unwrap_or(());
-                    process::exit(1);
-                }
-                Err(err) => {
-                    println!("Error: {}", err);
-                    process::exit(1);
-                }
+
+            let status = process::Command::new(&args.healthcheck).status()?;
+
+            if status.success() {
+                let message = [("WATCHDOG", "1")];
+                daemon::pid_notify(pid, false, message.iter())?;
             }
+
+            thread::sleep(interval);
         }
+    } else {
+        let pid = getpid();
+
+        // First we start the helper child process that will run the
+        // healthcheck
+        let mut helper = process::Command::new("/proc/self/exe")
+            .args(&["--healthcheck", &args.healthcheck])
+            .args(&["--pid", pid.to_string().as_str()])
+            .spawn()?;
+
+        // Then we execve to the requested program
+        let err = exec::Command::new(&args.free[0])
```

Comment:
This will panic if no free args are passed.

This is a  bug was already there, it was not introduced by the refactoring.

Still, I think it's a good moment to fix it and update the CLI help message to make clear that the command must be passed. I implemented this fix on this branch: https://github.com/balena-os/healthdog-rs/compare/majorz/refactoring...lmb/refactoring

Reply:
I cherry picked the commit.